hedera-node/src/main/java/com/hedera/services/fees/charging/NarratedCharging.java

```diff
@@ -20,21 +20,27 @@
  * ‍
  */
 
-import com.hederahashgraph.api.proto.java.AccountID;
+import com.hedera.services.utils.TxnAccessor;
+import com.hederahashgraph.fee.FeeObject;
 
 /**
- * Defines a type able to determine if a given payer can afford
- * to pay a given amount.
- *
- * @author Michael Tinker
+ * Defines the checks and charging actions we need to apply the Services fee policy.
  */
-@FunctionalInterface
-public interface BalanceCheck {
-	/**
-	 * Flags if the given payer can afford the given amount.
-	 * @param payer a payer
-	 * @param amount an amount
-	 * @return if the payer can afford the amount
-	 */
-	boolean canAfford(AccountID payer, long amount);
+public interface NarratedCharging {
+	void setFees(FeeObject fees);
+	void resetForTxn(TxnAccessor accessor, long submittingNodeId);
+
+	boolean canPayerAffordAllFees();
+	boolean canPayerAffordNetworkFee();
+	boolean canPayerAffordServiceFee();
+	boolean isPayerWillingToCoverAllFees();
+	boolean isPayerWillingToCoverNetworkFee();
+	boolean isPayerWillingToCoverServiceFee();
+
+	void chargePayerAllFees();
+	void chargePayerServiceFee();
+	void chargePayerNetworkAndUpToNodeFee();
```

Comment:
Shouldn't it be the other way around? I think we should charge payer up to the whole node fee first. Then if the offered fee could also cover the whole network fee, then we charge the network fee and submit the transaction. If not, we should not charge any partial network fee, and do not submit the transaction to the network.

Reply:
The network fees have priority (in general, the node receives its fee as a reward for satisfying the network preconditions; one of which is that the payer was both willing and able to pay the network fee).

Reply:
I have not thought of node fee as a reward. This makes sense.